### PR TITLE
Plugins

### PR DESCRIPTION
**Pull request by Nyholm**
Added plugins with the configuration feature discussed in #15.

### CODE REVIEW

File: DependencyInjection/HttplugExtension.php [hunk 1]
```diff
@@ -45,10 +57,22 @@ public function load(array $configs, ContainerBuilder $container)
             }
 
             $def = $container->register('httplug.client.'.$name, DummyClient::class);
-            $def->setFactory([new Reference($arguments['factory']), 'createClient'])
-                ->addArgument($arguments['config']);
+
+            if (empty($arguments['plugins'])) {
```
Comment (Nyholm) [inline DependencyInjection/HttplugExtension.php]: Do we want this? An alternative could be to send everything to PluginClientFactory so every client is an instance of PluginClient.

File: DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -90,6 +90,10 @@ protected function configureClients(ArrayNodeDefinition $root)
                         ->cannotBeEmpty()
                         ->info('The service id of a factory to use when creating the adapter.')
                     ->end()
+                    ->arrayNode('plugins')
+                        ->info('A list of service ids of plugins. The order is imortant.')
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: /s/imortant/important/

File: DependencyInjection/HttplugExtension.php (hunk 1 above)
Comment (dbu) [inline DependencyInjection/HttplugExtension.php]: i would keep it. there is no way to add plugins later at runtime, so it would only be relevant for people that typehint the concrete class instead of the interface, which would be so wrong that we should not care about helping them :-)

File: README.md [hunk 3]
```diff
@@ -99,6 +99,27 @@ $httpClient = $this->container->get('httplug.client.my_guzzle5');
...
 ```
 
+#### Plugins
+
+You can configure the clients with plugins. 
+
+```yaml
+// services.yml
+my_cache_plugin:
+  class: Http\Client\Plugin\CachePlugin
```
Comment (dbu) [inline README.md]: i think we should use a fictional plugin class name here, as we provide the standard plugins as services.

File: Resources/config/plugins.xml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <service id="httplug.plugin.cache" class="Http\Client\Plugin\CachePlugin" public="false">
+            <argument />
```
Comment (dbu) [inline Resources/config/plugins.xml]: where is this completed? do we need configuration for this?

File: Resources/config/plugins.xml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <service id="httplug.plugin.cache" class="Http\Client\Plugin\CachePlugin" public="false">
```
Comment (dbu) [inline Resources/config/plugins.xml]: agree that the plugins should be private (= removed from the container after it is built). we should list the services that are available in the documentation, and also remind that they are private and can thus not be manually fetched from the container.

**Comment by dbu** [comment]
:+1: thanks tobias!

File: Resources/config/plugins.xml (hunk 4 above)
Comment (Nyholm) [inline Resources/config/plugins.xml]: Hm, no it is not completed. I missed that. I have excluded a couple of plugin that require dependencies. The cache plugin must have a PSR6 cache pool and there is no way for me to give a good default. 

I think the best solution is to allow some sort of configuration like: 

```yaml
httplug: 
  plugins: 
    cache:
      service: 'acme.cache'
      config: 
        respect_headers: true
    logger:
      formatter: 'acme.formatter'
```

If you don't mind, could I do that in a separate PR?

**Comment by Nyholm** [comment]
Thank you for reviewing this so quickly. I've updated according to your comments.

File: Resources/config/plugins.xml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <service id="httplug.plugin.content_length" class="Http\Client\Plugin\ContentLengthPlugin" public="false" />
+        <service id="httplug.plugin.decoder" class="Http\Client\Plugin\DecoderPlugin" public="false">
+            <argument />
```
Comment (dbu) [inline Resources/config/plugins.xml]: there is an undefined argument here

File: Resources/config/plugins.xml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+        </service>
+        <service id="httplug.plugin.error" class="Http\Client\Plugin\ErrorPlugin" public="false" />
+        <service id="httplug.plugin.logger" class="Http\Client\Plugin\LoggerPlugin" public="false">
+            <argument type="service" id="logger"/>
+            <argument />
```
Comment (dbu) [inline Resources/config/plugins.xml]: and here as well

**Comment by dbu** [comment]
i agree solving plugin configuration in a separate PR. but right now there are two plugin configs with empty <argument/> tags, i think that will cause problems, no? if we can omit the params we should, otherwise we should remove those plugins for now.

File: Resources/config/plugins.xml (hunk 6 above)
Comment (Nyholm) [inline Resources/config/plugins.xml]: Isn't that okey since the decoder has a default value there?

**Comment by Nyholm** [comment]
We can omit those. 

Ignore my comment in: https://github.com/php-http/HttplugBundle/pull/16#discussion_r48978328

File: Resources/config/plugins.xml (hunk 6 above)
Comment (dbu) [inline Resources/config/plugins.xml]: not sure. do the tests cover this? if we have a fuctional test that manages
to load an actual client service with plugins set on it, we would see. atm
we lack a bit on testing afaik

Comment (Nyholm) [inline Resources/config/plugins.xml]: We do not have tests that cover this. You are correct that the bundle lacks tests.

**Comment by dbu** [comment]
alright, i think this is ready to merge. @Nyholm can you please squash the commits? feel free to merge, things look good for me.

**Comment by Nyholm** [comment]
Done. I've squashed my commints

**Comment by dbu** [comment]
great, thanks a lot!

**Comment by Nyholm** [comment]
Thank you for merging.
